src/hooks/usePipelineRuns.ts

```diff
@@ -0,0 +1,271 @@
+  const watchOptions = React.useMemo(() => {
+    // reset cached runs as the options have changed
+    etcdRunsRef.current = [];
+    return namespace
+      ? {
+          groupVersionKind,
+          namespace,
+          isList,
+          selector: optionsMemo?.selector,
+          name: optionsMemo?.name,
+        }
+      : null;
+  }, [groupVersionKind, namespace, optionsMemo, isList]);
+
+  const [resources, loaded, error] = useK8sWatchResource(watchOptions);
+  // if a pipeline run was removed from etcd, we want to still include it in the return value without re-querying tekton-results
+  const etcdRuns = React.useMemo(
+    () => (isList ? resources : loaded ? [resources] : []) as Kind[],
+    [isList, resources, loaded],
+  );
+
+  const runs = React.useMemo(() => {
+    if (!etcdRuns) {
+      return etcdRuns;
+    }
+    let value = etcdRunsRef.current
+      ? [
+          ...etcdRuns,
+          // identify the runs that were removed
+          ...differenceBy(etcdRunsRef.current, etcdRuns, (plr) => plr.metadata.name),
+        ]
+      : [...etcdRuns];
+    value.sort((a, b) => b.metadata.creationTimestamp.localeCompare(a.metadata.creationTimestamp));
+    if (limit && limit < value.length) {
+      value = value.slice(0, limit);
+    }
+    return value;
+  }, [etcdRuns, limit]);
+
+  // cache the last set to identify removed runs
+  etcdRunsRef.current = runs;
+
+  // Query tekton results if there's no limit or we received less items from etcd than the current limit
+  const queryTr =
+    !limit || (namespace && ((runs && loaded && optionsMemo.limit - runs.length > 0) || error));
+
+  const trOptions: typeof optionsMemo = React.useMemo(() => {
+    if (optionsMemo?.name) {
+      const { name, ...rest } = optionsMemo;
+      return {
+        ...rest,
+        filter: EQ('data.metadata.name', name),
+      };
+    }
+    return optionsMemo;
+  }, [optionsMemo]);
+
+  // tekton-results includes items in etcd, therefore options must use the same limit
+  // these duplicates will later be de-duped
+  const [trResources, trLoaded, trError, trGetNextPage] = (
+    groupVersionKind === PipelineRunGroupVersionKind ? useTRPipelineRuns : useTRTaskRuns
+  )(queryTr ? namespace : null, trOptions) as [Kind[], boolean, unknown, GetNextPage];
+
+  return React.useMemo(() => {
+    const rResources =
+      runs && trResources
+        ? uniqBy([...runs, ...trResources], (r) => r.metadata.name)
+        : runs || trResources;
+
+    return [
+      rResources,
+      namespace
+        ? queryTr
+          ? isList
+            ? // return loaded only if both sources have loaded
+              trLoaded && loaded
+            : // when searching by name, loading fails if we have no result
+              !!rResources?.[0]
+          : isList
+          ? loaded
+          : // when searching by name, loading fails if we have no result
+            !!rResources?.[0]
+        : false,
```

Comment:
Should this be `loaded`?

Reply:
we return `false` when there is no `namespace`